crates/red_knot_test/src/parser.rs

```diff
@@ -381,19 +387,35 @@ impl<'s> Parser<'s> {
             return self.parse_config(code);
         }
 
+        let explicit_path = config.get("path").copied();
+
+        if let Some(explicit_path) = explicit_path {
+            if !lang.is_empty()
+                && explicit_path.contains('.')
+                && !explicit_path.ends_with(&format!(".{lang}"))
+            {
+                return Err(anyhow::anyhow!(
+                    "File `{explicit_path}` has mismatching `lang={lang}`"
+                ));
+            }
+        }
+
+        let path = explicit_path.unwrap_or(match lang {
+            "pyi" => "test.pyi",
+            _ => "test.py",
+        });
+
         self.files.push(EmbeddedFile {
             path,
             section,
             lang,
-
             code,
-
             md_offset: self.offset(),
         });
 
         if let Some(current_files) = &mut self.current_section_files {
             if !current_files.insert(path) {
-                if path == "test.py" {
+                if matches!(path, "test.py" | "test.pyi") {
                     return Err(anyhow::anyhow!(
                         "Test `{}` has duplicate files named `{path}`. \
                                 (This is the default filename; \
```

Comment:
On the line immediately following this, we should update the error message to recommend the new-style way of giving the file path instead.